### PR TITLE
Propagate Parties.t failures to apply_transaction

### LINKED ISSUE #9592: snapp transaction failure cases

### PR DESCRIPTION
**Pull request by psteckler**
Propagate failures in `Parties_logic` to errors in `Transaction_logic.apply_transaction`.

The strategy for that is to add a new field `failure_status` in `Parties_logic.Local_state.t`. Outside the transaction SNARK, that's a `Transaction_status.Failure.t option`; inside the SNARK, it's `unit`.

In `apply_parties_unchecked_aux`, if a step returns a local status with a failure, we convert that failure to an error.

Some bits to review carefully: 
- `Mina_state.Local_state.to_input` changes, which relies on `Transaction_status.Failure.option_to_input`
- in the same module, `failure_status_typ`, which relates `failure_status` types in/out the SNARK
- also `gen` in that module, which relies on `Transaction_status.Failure.gen`

Closes #9592

### CODE REVIEW

File: src/lib/mina_base/parties_logic.ml [hunk 1]
```diff
@@ -1,3 +1,5 @@
+(* parties_logic.ml *)
```
Comment (mrmr1993) [inline src/lib/mina_base/parties_logic.ml:1]: Still not a fan of this style :)

File: src/lib/mina_state/local_state.ml [hunk 2]
```diff
@@ -133,6 +155,16 @@ module Checked = struct
       |]
 end
 
+(* there: map any failure status to the unit value in Checked
+   back: map the unit value in Checked to None in the value world
+   (an alternative would be to fail, since we intend never to do that)
```
Comment (mrmr1993) [inline src/lib/mina_state/local_state.ml:160]: Worth saying: we don't do this to increase ease of debugging?

File: src/lib/mina_base/parties_logic.ml [hunk 3]
```diff
@@ -374,6 +387,7 @@ module Make (Inputs : Inputs_intf) = struct
          ; transaction_commitment : Transaction_commitment.t
          ; amount : Amount.t
          ; bool : Bool.t
+         ; failure : Transaction_status.Failure.t option
```
Comment (mrmr1993) [inline src/lib/mina_base/parties_logic.ml:390]: Nit: make this a variable, like `global_state`

File: src/lib/transaction_snark/transaction_snark.ml [hunk 4]
```diff
@@ -1813,11 +1813,13 @@ module Base = struct
...
               , Bool.t
-              , Transaction_commitment.t )
+              , Transaction_commitment.t
+              , unit )
               Parties_logic.Local_state.t
           ; protocol_state_predicate : Snapp_predicate.Protocol_state.Checked.t
           ; transaction_commitment : Transaction_commitment.t
-          ; field : Field.t >
+          ; field : Field.t
+          ; failure : Transaction_status.Failure.t option >
```
Comment (mrmr1993) [inline src/lib/transaction_snark/transaction_snark.ml:1822]: Nit: `unit` here?

File: src/lib/mina_base/transaction_status.ml [hunk 5]
```diff
@@ -106,6 +114,31 @@ module Failure = struct
     | _ ->
         Error "Signed_command_status.Failure.of_string: Unknown value"
 
+  let to_input, option_to_input =
```
Comment (mrmr1993) [inline src/lib/mina_base/transaction_status.ml:117]: Nit: delete

**Review by mrmr1993 — APPROVED** [review]
Requested some nit-ty changes. Happy for this to merge past the versioning gate, changes look safe.

File: src/lib/mina_base/parties_logic.ml (hunk 1 above)
Comment (psteckler) [reply src/lib/mina_base/parties_logic.ml:1]: This style comes from a textbook I used years ago. It's helpful if you ever print out a file. I guess people don't do that so much these days.

**Review by mrmr1993 — APPROVED** [review]

**Pull request merged**